flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogCleanerImpl.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+class FsStateChangelogCleanerImpl implements FsStateChangelogCleaner {
+    private static final Logger LOG = LoggerFactory.getLogger(FsStateChangelogCleanerImpl.class);
+    private static final int MAX_TASKS_PER_THREAD = 100;
+
+    private final Executor executor;
+
+    public FsStateChangelogCleanerImpl(int nThreads) {
+        // Use a fixed-size thread pool with a bounded queue so that cleanupAsync back pressures
+        // callers if the cleanup doesn't keep up.
+        // In all cases except abort this are uploader threads; while on abort this is the task
+        // thread.
+        this(
+                new ThreadPoolExecutor(
+                        nThreads,
+                        nThreads,
+                        0L,
+                        TimeUnit.MILLISECONDS,
+                        new LinkedBlockingQueue<>(nThreads * MAX_TASKS_PER_THREAD),
+                        (ThreadFactory) Thread::new));
```

Comment:
Let's not create yet another thread pool. `StreamTask#asyncOperationsThreadPool` sounds like a pool that is already doing IO against the same external system.

Reply:
Could you explain why not?

I see the following reasons to use a separate pool:
1. (as mentioned in the comment): use a blocking queue to provide back-pressure
2. It's easier to configure size with separate pools
3. Easier to debug

I'm assuming the overhead of having more threads per task bounded by `nThreads` negligible on modern OSes.

Reply:
I'm not worried about the overhead of having an idle thread pool.

1. I'm not sure if it's currently implemented, but having a backpressure on {{AsyncCheckpointRunnable}} also sounds reasonable.
2. If we have multiple pools doing very similar things, that should be tweaked in very similar way it's not easier to configure all of them independently for the user. At best user would need to know about each of those pools, and even assuming that he would have a problem if he wants to limits parallel threads writing to his DFS to for example 4, regardless if that's {{AsyncCheckpointRunnable}}, periodic materialisation or changelog writes.
3. Is it? What if you want to check if number of threads is an issue and you just simply want to bump/decrease the number of threads by a factor of 4? With single pool it's easier to do. With many pools you might hit a problem that you simply forgotten to modify that one thread pool that was actually causing the problem.

Reply:
I looked at the code again and I think that async cleanup is actually not necessary in this PR. It only happens on shutdown (and in that case it could be done by uploader threads).
In the previous PR it was called in more cases because of aborts.

So I'm going to remove it alltogether.